gcn_classic_text_to_json/utils.py

```diff
@@ -0,0 +1,192 @@
+def find_string_between(data, first, last, start_idx):
+    """Find string between `first` and `last` beginning at `start_idx`.
+
+    Parameters
+    ----------
+    data: string
+        The string through which to search for `first` and `last`.
+    first: string
+        The text preceeding the string to be returned.
+    last: string
+        The text succeeding the string to be returned.
+    start_idx: int
+        The index from which to begin search for `first`.
+
+    Returns
+    -------
+    string
+        The text between `first` and `last`."""
+    first_idx = data.find(first, start_idx) + len(first)
+    last_idx = data.find(last, first_idx)
+    return data[first_idx:last_idx]
+
+
+def find_notice_keyword(data, notice_keyword, notice_start_idx):
+    """Finds the data associated with `notice_keyword` in the text notices.
```

Comment:
A lot of the text based alerts can be parsed as email headers. Python's email package has a couple functions for parsing these out, https://docs.python.org/3/library/email.parser.html
In this case, message_from_string might work for you. You could potentially take a whole alert as a string, and do something like:

```python
import email

# Stuff to get the notice as a string...

keyword_dict = dict(email.message_from_string(notice))
```

Note, you may have to trim some whitespace here and there

Reply:
What did `data[start_idx:end_idx]` look like when you printed them?

Reply:
```

TITLE:           GCN/CALET NOTICE
NOTICE_DATE:     Fri 28 Jun 24 16:28:49 UT
NOTICE_TYPE:     CALET GBM Flight Lightcurve 
POINT_RA:        293.30d {+19h 33m 12s} (J2000),
POINT_DEC:       +35.98d {+35d 58' 59"} (J2000),
TRIGGER_NUM:     1403626815
TRIGGER_DATE:    20489 TJD;   180 DOY;   24/06/28
TRIGGER_TIME:    59295.28 SOD {16:28:15.28} UT
TRIGGER_DET:     1 1 1 (HXM1 HXM2 SGM)
SIGNIFICANCE:    15.69 [sigma]
FOREGND_DUR:     1.00 [sec]
BACKGND_DUR1:    16.00 [sec]
ENERGY_BAND:     40-230 [keV]
SC_LON_LAT:      129.19,36.02 [deg]
LC_URL:          http://cgbm.calet.jp/cgbm_trigger/flight//1403626815/index.html
COMMENTS:        CALET-GBM Flight Lightcurve.  
COMMENTS:        The LC_URL file will not be created until ~15 min after the trigger.  
 
 
```

Reply:
Try calling `.lstrip()` to trim off leading whitespace from `data[start_idx:end_idx]`:

```python
temp = """
TITLE:           GCN/CALET NOTICE
NOTICE_DATE:     Fri 28 Jun 24 16:28:49 UT
NOTICE_TYPE:     CALET GBM Flight Lightcurve
POINT_RA:        293.30d {+19h 33m 12s} (J2000),
POINT_DEC:       +35.98d {+35d 58' 59"} (J2000),
TRIGGER_NUM:     1403626815
TRIGGER_DATE:    20489 TJD;   180 DOY;   24/06/28
TRIGGER_TIME:    59295.28 SOD {16:28:15.28} UT
TRIGGER_DET:     1 1 1 (HXM1 HXM2 SGM)
SIGNIFICANCE:    15.69 [sigma]
FOREGND_DUR:     1.00 [sec]
BACKGND_DUR1:    16.00 [sec]
ENERGY_BAND:     40-230 [keV]
SC_LON_LAT:      129.19,36.02 [deg]
LC_URL:          http://cgbm.calet.jp/cgbm_trigger/flight//1403626815/index.html
COMMENTS:        CALET-GBM Flight Lightcurve.
COMMENTS:        The LC_URL file will not be created until ~15 min after the trigger.

 """
d = dict(email.message_from_string(temp))
d
{}
d = dict(email.message_from_string(temp.lstrip()))
d
{'TITLE': 'GCN/CALET NOTICE', 'NOTICE_DATE': 'Fri 28 Jun 24 16:28:49 UT', 'NOTICE_TYPE': 'CALET GBM Flight Lightcurve', 'POINT_RA': '293.30d {+19h 33m 12s} (J2000),', 'POINT_DEC': '+35.98d {+35d 58\' 59"} (J2000),', 'TRIGGER_NUM': '1403626815', 'TRIGGER_DATE': '20489 TJD;   180 DOY;   24/06/28', 'TRIGGER_TIME': '59295.28 SOD {16:28:15.28} UT', 'TRIGGER_DET': '1 1 1 (HXM1 HXM2 SGM)', 'SIGNIFICANCE': '15.69 [sigma]', 'FOREGND_DUR': '1.00 [sec]', 'BACKGND_DUR1': '16.00 [sec]', 'ENERGY_BAND': '40-230 [keV]', 'SC_LON_LAT': '129.19,36.02 [deg]', 'LC_URL': 'http://cgbm.calet.jp/cgbm_trigger/flight//1403626815/index.html', 'COMMENTS': 'CALET-GBM Flight Lightcurve.'}

```

Though you may see issues when there are multiple COMMENTS fields, so if this doesn't prove helpful for you, feel free to ignore it and continue as you were

Reply:
Note, if you don't immediately convert to a dict, you can also use the parsed message's `get_all` method to retrieve all the comments so as not to lose them. You will see it in the keys more than once:

```python 

e = email.message_from_string(temp.lstrip())
e.keys()
# ['TITLE', 'NOTICE_DATE', 'NOTICE_TYPE', 'POINT_RA', 'POINT_DEC', 'TRIGGER_NUM', 'TRIGGER_DATE', 'TRIGGER_TIME', 'TRIGGER_DET', 'SIGNIFICANCE', 'FOREGND_DUR', 'BACKGND_DUR1', 'ENERGY_BAND', 'SC_LON_LAT', 'LC_URL', 'COMMENTS', 'COMMENTS']
e.get_all('COMMENTS')
# ['CALET-GBM Flight Lightcurve.', 'The LC_URL file will not be created until ~15 min after the trigger.']
```